### PR TITLE
Put version information in separate header

### PR DESCRIPTION
**Pull request by floli**
Related to #117

### CODE REVIEW

**Comment by floli** [comment]
Can't be merged yet, because it fails with scons.

File: CMakeLists.txt [hunk 1]
```diff
@@ -39,6 +30,14 @@ endif()
 
 option(BUILD_SHARED_LIBS "Build a shared library" ON)
 
+# Get preCICE version from git (that assumes you are compiling on a git checkout)
```
Comment (fsimonis) [inline CMakeLists.txt]: Either assert this assumption or handle the edge case.
Distributing archived source code without `.git` is a common example that does not satisfy this assumption.

File: CMakeLists.txt [hunk 2]
```diff
@@ -99,6 +96,10 @@ else()
 endif()
 
 
+# Create versions.hpp and add binary dir to include dirs
+configure_file("${PROJECT_SOURCE_DIR}/src/versions.hpp.in" "${PROJECT_BINARY_DIR}/versions.hpp")
```
Comment (fsimonis) [inline CMakeLists.txt]: Please restrict substitution using `@ONLY` to enforce consistency in the future.

https://cmake.org/cmake/help/latest/command/configure_file.html

**Comment by floli** [comment]
Just using a plain version number, e.g. 1.2.0 is fine for released version, however, if someone uses a non-released version, it's nicer to also have the commit id.

**Comment by MakisH** [comment]
For the xSDK compliance we need to have a version number for the released versions and a commit id for the development branch. I agree that we should not assume git always.

I still haven't found time to review this, but I will!

**Comment by fsimonis** [comment]
@MakisH does the xSDK specify a version number rule for modified non-git repos?

If that is not the case, we could:

1) try to FindGit
2) try to get the commit (the source might not be a git repo)
3) try to detect a modified build (no tag?)
4) __then__ set the version to the commit hash

If __any__ of these steps fail, we could simply assume that the project is a unmodified version.

We would violate the [cmake version format](https://cmake.org/cmake/help/latest/command/project.html) though:
> If VERSION is specified, given components must be non-negative integers. If VERSION is not specified, the default version is the empty string.

**Comment by MakisH** [comment]
Here is the exact description from xSDK:

> M8.​​ ​Each​ ​package​ ​must​ ​provide​ ​a​ ​runtime​ ​API​,​ ​for​ ​example​ ​a​ ​function​ ​call,​ ​to​ ​return​ ​the​ ​current version​ ​number​​ ​of​ ​the​ ​software.​ ​ ​For​ ​development​ ​versions​ ​of​ ​the​ ​software,​ ​each​ ​package​ ​must provide​ ​the​ ​current​ ​commit​ ​ID​ ​in​ ​the​ ​repository​ ​and​ ​indicate​ ​what​ ​compiler​ ​options​ ​were​ ​used​ ​to​ ​build the​ ​package.​ ​With​ ​this​ ​information​ ​users​ ​should​ ​be​ ​able​ ​to​ ​rebuild​ ​the​ ​package​ ​in​ ​the​ ​same​ ​state.​ ​We do​ ​not​ ​currently​ ​require​ ​that​ ​version​ ​information​ ​for​ ​all​ ​dependent​ ​packages​ ​be​ ​provided,​ ​so​ ​it​ ​may​ ​not be​ ​possible​ ​to​ ​rebuild​ ​the​ ​entire​ ​software​ ​stack​ ​in​ ​the​ ​same​ ​state.

The important message here is to have a way to refer to and reproduce specific states of the code.

**Comment by fsimonis** [comment]
Ok, so lets fiddle this apart:

> Each​ ​package​ ​must​ ​provide​ ​a​ ​runtime​ ​API​,​ ​for​ ​example​ ​a​ ​function​ ​call,​ ​to​ ​return​ ​the​ ​current version​ ​number​​ ​of​ ​the​ ​software.​ 

So, we need to provide runtime access to this information.
As we are using C++11, we could do the following:
```cpp
namespace precice { namespace version {
    static constexpr char * version = "@PRECICE_VERSION@";
    static constexpr uint8_t version_major = @PRECICE_VERSION_MAJOR@;
    static constexpr uint8_t version_minor = @PRECICE_VERSION_MINOR@;
    static constexpr uint8_t version_patch = @PRECICE_VERSION_PATCH@;
}}
```

> ​For​ ​development​ ​versions​ ​of​ ​the​ ​software,​ ​each​ ​package​ ​must provide​ ​the​ ​current​ ​commit​ ​ID​ ​in​ ​the​ ​repository

This reads like it simply needs to be available somewhere.
We could generate a simple file `commit.txt` via cmake only containing the commit hash and install it along with the package.
Alternatively, we could try to incorporate this into the _hopefully future_ `preciceConfig.cmake`.

>​ ​and​ ​indicate​ ​what​ ​compiler​ ​options​ ​were​ ​used​ ​to​ ​build the​ ​package.​ ​With​ ​this​ ​information​ ​users​ ​should​ ​be​ ​able​ ​to​ ​rebuild​ ​the​ ​package​ ​in​ ​the​ ​same​ ​state.​ ​We do​ ​not​ ​currently​ ​require​ ​that​ ​version​ ​information​ ​for​ ​all​ ​dependent​ ​packages​ ​be​ ​provided,​ ​so​ ​it​ ​may​ ​not be​ ​possible​ ​to​ ​rebuild​ ​the​ ​entire​ ​software​ ​stack​ ​in​ ​the​ ​same​ ​state.

The overkill version of this is already available in CMake by telling it to [export the compile commands](https://cmake.org/cmake/help/latest/variable/CMAKE_EXPORT_COMPILE_COMMANDS.html).
The resulting JSON contains the compile command for each translation units, so we could put that in the package too.

**Comment by floli** [comment]
Addendum: `constexpr` are nicer than `#define` but you can't use them for conditional compilation before C++17.

**Pull request merged**